### ISSUE TITLE
v5 - TypeError: In this environment the sources for assign MUST be an object.This error is a performance optimization and not spec compliant

### ISSUE THREAD

**Issue by ranggarifqi** [rt2zz/redux-persist#536]
Hey, umm... i've tried to implement redux persist for my project...
After i followed your documentation on how to implement this module, there's no problem.

but, when i tried to reload the app...Then it shows a blank white screen and shows warning `TypeError: In this environment the sources for assign MUST be an object.This error is a performance optimization and not spec compliant`

Then, i tried to change the config key from `root` to anything... like `rootzzz`. And it actually works. But then when i tried to change it again to `root`, it shows this blank screen again, with those warning

I don't understand what's going on... could you help me ?

this is my reducer list `./src/reducers/index.js`

```
import {combineReducers} from 'redux';
import { persistCombineReducers } from 'redux-persist'
import storage from 'redux-persist/es/storage';

const config = {
  key: 'root',
  storage,
}

import { reducer as formReducer } from 'redux-form'
import AuthReducer from './AuthReducer';
import SideMenuReducer from './SideMenuReducer';
import SideMenuSelectReducer from './SideMenuSelectReducer';
import HomeDashReducer from './HomeDashReducer';

const reducers = {
    form: formReducer,
    auth : AuthReducer,
    sidemenus: SideMenuReducer,
    selectedMenuId : SideMenuSelectReducer,
    menuHome : HomeDashReducer
}
export default persistCombineReducers(config, reducers);
```

and this is my code for `./App.js`

```
import React, {Component} from 'react';
import {View, AsyncStorage} from 'react-native';
import expo from 'expo';
import {Provider} from 'react-redux';
import {createStore, applyMiddleware} from 'redux';
import ReduxThunk from 'redux-thunk';
import reducers from './src/reducers';
import Router from './src/Router';

import { persistStore } from 'redux-persist';
import { PersistGate } from 'redux-persist/es/integration/react';

function configureStore () {
  // ...
  let store = createStore(reducers, {}, applyMiddleware(ReduxThunk))
  let persistor = persistStore(store)

  return { persistor, store }
}

class App extends Component{

    async componentWillMount(){
        expo.Font.loadAsync({
          'Roboto_medium': require('./assets/fonts/Roboto_medium.ttf'),
        });
    }

    onBeforeLift = () => {
      // take some action before the gate lifts
    }

    render(){
        const { store, persistor } = configureStore();
        return(

            <Provider store={store}>
                <PersistGate
                    loading={<View />}
                    onBeforeLift={this.onBeforeLift}
                    persistor={persistor}
                >
                    <Router/>
              </PersistGate>

            </Provider>
        );
    }
}

export default App;
```

Is there anything wrong with my code ?
its been 3 hours and i still have no clue

Thanks :)

**Comment by cheneywan** [comment]
The same problem..

**Comment by Cariosvertel** [comment]
Some problem too. FYI, you can customize blank screen by modifying this line `loading={<View />}`. I know it's something trivial because it's detailed on doc, but I known it could be helpful for someone who read this post.

**Comment by sun2rise** [comment]
Same issue here.. the only difference is that I use AsyncStorage (that should be the proper choice with RN)

```js
import { AsyncStorage } from 'react-native';

const config = {
  key: 'root',
  storage: AsyncStorage
};
```

**Comment by Cariosvertel** [comment]
according to [this line](https://github.com/rt2zz/redux-persist/blob/master/src/persistStore.js#L101) where error is fired, it should accept an object as `payload`, and seems that just stored data isn't being provided as it.

**Comment by sun2rise** [comment]
@ranggarifqi , @Cariosvertel .. it all started working after a complete uninstall of the app and a new deploy with `react-native run-android`.
Maybe the store was sill there but in a v4 fashion and everything, that's why you need to migrate or completely reinstall the app

**Comment by Cariosvertel** [comment]
@sun2rise I tried this line `watchman watch-del-all && rm -rf $TMPDIR/react-* && rm -rf node_modules/ && npm cache clean && npm install && npm start -- --reset-cache
` in order to reset me project but no success.

**Comment by jmcol** [comment]
I'm having the same problem, as well. I reinstalled my client simulator and wiped all my cache. That helped the simulator start working again; however, once I attempted to persist data, I got the error again.

**Comment by rkleine** [comment]
I was having the same issue. This maybe is because you have a state with a **root** key that the value is not an object, for example:

```js
{
  posts: [], // this is ok, is an object
  agreeToC: false // this is causing the issue because false can be deserialized
}
```

When trying to serialize the above object will throw the error. Try to change to samething like this

```js
{
  posts: [], // this is ok, is an object.
  agreeToC: { accept: false } // this will fix the issue
}
```

If you already have a persiste store with the error:

1.  Add this line `persistor.purge();` after `const persistor = persistStore(store);`
2. Reload the app so the store is purged, 
3. Remove/comment the line `persistor.purge();`
4. Reload the app and see if the issue is fixed

**Comment by ranggarifqi** [comment]
i ended up using v4 though
because of my deadline

Before i decided to use v4,
i think this error came from `PersistGate`

i tried to remove `PersistGate` and i didnt see the error. But, idk somehow it didnt rehydrate

Well, i'll try to tweak this again if i have time.

**Comment by rkleine** [comment]
@ranggarifqi do you try my suggestion?

**Comment by rt2zz** [comment]
that is correct, right now reducer state needs to be objects. I plan to update this in the near future to  support non-objects.

**Comment by rt2zz** [comment]
actually just cut redux-persist@5.3.0-rc which hopefully resolves this. `npm i redux-persist@next` to test

**Comment by sachin-sable** [comment]
@rt2zz I have combination of objects and non-objects. I am planning to update to v5 but this is the only thing that is stopping me as I am getting the same error stated above. I am using v5.4.0. Is there something I am missing to add or is my approach correct? Please help.

**Comment by rt2zz** [comment]
if you have non-object state then you will want to use persistReducer rather than persistCombineReducers.

```js
let rootReducer = combineReducers(reducers)
let persistedReducer = persistReducer(config, rootReducer)
```

**Comment by sachin-sable** [comment]
@rt2zz Thank you so much for help. I ended up converting all non object state to objects by grouping related items together. For e.g. globalLoading, error, isConnected, etc are grouped under object called "networkState". Thanks.

**Comment by rt2zz** [comment]
ya to elaborate a little bit, persistReducer only shallow merges 1 level or state, while persistCombineReducers merges 2. 

Still looking for how to surface these options in a developer friendly way... but ya for now persistReducer should be the way to go for your case 👍

**Comment by rochapablo** [comment]
I tried to change to 

```
// const reducer = persistCombineReducers(config, reducers)
let rootReducer = combineReducers(reducers)
let reducer = persistReducer(config, rootReducer)
```

but it keeps getting 

```
ExceptionsManager.js:73 Notification handler err TypeError: In this environment the sources for assign MUST be an object.This error is a performance optimization and not spec compliant.
    at Object.assign (Object.es6.js:42)
```

**Comment by DalalAbadi** [comment]
I am using firebase and react-redux and I solved this issue by adding this line 


`import _ from 'lodash';`

**Comment by chrisbull** [comment]
I get this error with `redux-persist@5.7.2` when I import autoMergeLevel2: 
`import autoMergeLevel2 from 'redux-persist/lib/stateReconciler/autoMergeLevel2'`

**Comment by rt2zz** [comment]
@chrisbull are you using immutablejs?
